### PR TITLE
[POC] build: Hello Qt 6

### LINKED ISSUE #20627: RFC: Migrate from Qt5 to Qt6's Qt5 compat
It seems like migrating to Qt6 is a win for multiple efforts, and I want us to evaluate switching to it as an alternative to monkey-patching Qt5/qmake to make it work properly.

1. First class support for Apple Silicon for #20371 (see: https://bugreports.qt.io/browse/QTBUG-85279)
2. Transition from `qmake` to `cmake` (see: https://www.qt.io/blog/qt-6-build-system)
	1. We already use `cmake` in depends to build several other tools
	2. `cmake` is much better maintained than `qmake`
	3. This might make a native/build split possible for `qt`, the last holdout package of native/build mixing after `boost` was recently split
	4. For PRs like #20600, where we're trying to work around `qmake`'s jankiness, I don't want people to pour their heart and soul into making `qmake` work just to be completely replaced by `cmake` when we bump in the future

Supposedly, the switching cost is also minimal, as there's a Qt5Compat module (see: https://www.qt.io/blog/porting-from-qt-5-to-qt-6-using-qt5compat-library)

### PR DESCRIPTION
**Pull request by hebasto**
This is a draft of upgrading Qt in depends up to [Qt 6.2 LTS](https://www.qt.io/blog/qt-6.2-lts-released).

The motivation is:
- Qt 6 seems (almost) inevitable in the future
- this change drops all of the backported patches (but introduces a new one)

Tasks
- [x] backport the QTBUG-86080 [fix](https://github.com/qt/qtbase/commit/5b07f14a4ff6b08887473b1b583ee2f71dec6f76)
- [x] integrate Qt 6 into `build-aux/m4/bitcoin_qt.m4`
- [x] fix builds which are configured with `--disable-wallet`
- [ ] cross-compiling
- [ ] Guix builds

Closes bitcoin/bitcoin#20627.

---

Steps to build on Linux x86_64:
```
$ make -C depends
$ ./autogen.sh
$ ./configure --with-gui=qt6 CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site
$ make clean 
$ make
$ make check
```

### CODE REVIEW

**Comment by laanwj** [comment]
So I understand this doesn't need any actual code changes? Only build system and depends?

**Comment by hebasto** [comment]
> So I understand this doesn't need any actual code changes? Only build system and depends?

Cannot say for sure at this point.

**Comment by laanwj** [comment]
> Cannot say for sure at this point.

OK I assumed so, as it wouldn't make much sense to change depends if the code cannot handle it. We could try compiling against system Qt6.

**Comment by hebasto** [comment]
> > Cannot say for sure at this point.
> 
> OK I assumed so, as it wouldn't make much sense to change depends if the code cannot handle it. We could try compiling against system Qt6.

Some changes are required: bitcoin-core/gui#577, bitcoin-core/gui#578.

**Comment by hebasto** [comment]
Compiled and linked (not pushed yet):
```
2022-04-09T11:25:38Z Bitcoin Core version v23.99.0-388ee2edb127 (release build)
2022-04-09T11:25:38Z Qt 6.2.4 (static), plugin=xcb (static)
2022-04-09T11:25:38Z Static plugins:
2022-04-09T11:25:38Z  QXcbIntegrationPlugin, version 393728
2022-04-09T11:25:38Z Style: fusion / QFusionStyle
2022-04-09T11:25:38Z System: Ubuntu Jammy Jellyfish (development branch), x86_64-little_endian-lp64
...
```

Screenshot from 2022-04-09 13-24-24

**Comment by hebasto** [comment]
Steps to build this PR in its current state on Linux x86_64:
```
$ make -C depends
$ ./autogen.sh
$ CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site ./configure
$ make clean 
$ make
```

**Comment by hebasto** [comment]
I ran into a problem while working on this PR.

The problem's root is in two facts:
1) our build system rely on `pkg-config`
2) Qt 6 build system, after switching from `qmake` to `cmake`, does not produce `*.pc` files

The fix suggested in [QTBUG-86080](https://bugreports.qt.io/browse/QTBUG-86080) works only for _shared_ libraries.

Here are some ways to work out this problem:
1) modify Qt's patch to support static builds. Nevertheless, the generated `*.pc` files lack `*.private` sections which contains libraries required for static builds. This issue in turn can be solved by adjusting resulted `*_LIB` variables ad-hoc. Btw, this approach is currently implemented in this PR
2) develop our own patch which make Qt generate correct `*.pc` files for static builds (the content of internal `QMAKE_PRL_LIBS` variables could be used for that)
3) keep working `*.pc` files in our repo, and substitute them for static  builds
4) something else
5) ...

**Comment by jarolrod** [comment]
While it's good to think about this kind of integration, I don't know how feasible this will ultimately be. 

In [d0015fd5f3b0465d2b3f36fda3714e248f1b4c51](https://github.com/bitcoin/bitcoin/pull/24798/commits/d0015fd5f3b0465d2b3f36fda3714e248f1b4c51), you uncomment lines where the patch author intended to return when performing a static build. Here's a [link](https://bugreports.qt.io/browse/QTBUG-86080?focusedCommentId=649261&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-649261) to a discussion on the [QTBUG-86080](https://bugreports.qt.io/browse/QTBUG-86080) page where it talks about the current qt internal issues with `*.pc` files and static builds.

On the surface, it appears we most likely can get this working. But, I worry about the implications of the hacky path we'd take to get there.

**Comment by laanwj** [comment]
IMO for static linking in the release, it's acceptable to carry some hacky patches in depends until the issue gets resolved upstream. I don't have any specific opinion which one, as long as it's works and is maintainable.

I do think we should be able to build against distro Qt6 -dev packages for local builds for development. But dynamic-only is fine there.

**Comment by luke-jr** [comment]
Would prefer to see this PR split between Qt6 compatibility (goal ASAP), and using Qt6 for releases (goal TBD later).

**Comment by hebasto** [comment]
> Would prefer to see this PR split between Qt6 compatibility (goal ASAP), and using Qt6 for releases (goal TBD later).

https://github.com/bitcoin/bitcoin/pull/24798#issuecomment-1093493137 -- is it what you mean?

**Comment by hebasto** [comment]
Rebased on top of the bitcoin-core/gui#577.

**Comment by hebasto** [comment]
Rebased on top of the bitcoin-core/gui#579.

**Comment by hebasto** [comment]
Rebased on top of the bitcoin-core/gui#580 and bitcoin-core/gui#584.

File: build-aux/m4/bitcoin_qt.m4 [hunk 1]
```diff
@@ -372,16 +411,34 @@ dnl Outputs: All necessary QT_* variables are set.
...
+    QT_INCLUDES="$QT_GUI_CFLAGS $QT_INCLUDES"
+    QT_LIBS="$QT_GUI_LIBS $QT_LIBS"
+    _BITCOIN_QT_CHECK_APP([QGuiApplication])
   ])
   BITCOIN_QT_CHECK([
-    PKG_CHECK_MODULES([QT_WIDGETS], [${qt_lib_prefix}Widgets${qt_lib_suffix} $qt_version], [QT_INCLUDES="$QT_WIDGETS_CFLAGS $QT_INCLUDES" QT_LIBS="$QT_WIDGETS_LIBS $QT_LIBS"],
+    PKG_CHECK_MODULES([QT_WIDGETS], [${qt_lib_prefix}Widgets${qt_lib_suffix} $qt_version], [],
                       [BITCOIN_QT_FAIL([${qt_lib_prefix}Widgets${qt_lib_suffix} $qt_version not found])])
+    if test "$bitcoin_qt_want_version" = "qt6"; then
+      QT_WIDGETS_LIBS="$QT_WIDGETS_LIBS ${qt_lib_path}/objects-Release/Widgets_resources_1/.rcc/qrc_qstyle.cpp.o ${qt_lib_path}/objects-Release/Widgets_resources_2/.rcc/qrc_qstyle1.cpp.o ${qt_lib_path}/objects-Release/Widgets_resources_3/.rcc/qrc_qmessagebox.cpp.o"
```
Comment (fanquake) [inline build-aux/m4/bitcoin_qt.m4:437]: In [9f61d6a7605eee4b116b8791276bd71baf75bf10](https://github.com/bitcoin/bitcoin/pull/24798/commits/9f61d6a7605eee4b116b8791276bd71baf75bf10):
Can you explain what's going on here? Trying to link against `.o` files in configure checks not only seems very fragile, but downright hacky.

Comment (hebasto) [reply build-aux/m4/bitcoin_qt.m4:437]: This is reproducing of lacking `pkg-config --static` functionality. List of files to link against has been taken from `*.prl` files.

**Comment by hebasto** [comment]
Rebased on top of the bitcoin-core/gui#589.

**Comment by hebasto** [comment]
Rebased on top of the bitcoin-core/gui#586.

**Comment by hebasto** [comment]
Fixed builds which are configured with `--disable-wallet`.

**Comment by fanquake** [comment]
Can you rebase this to reflect the current state of / remaining changes needed for integration? Seems that some of the changes here have now made it in (via the GUI repo).

**Comment by hebasto** [comment]
> Can you rebase this to reflect the current state of / remaining changes needed for integration? Seems that some of the changes here have now made it in (via the GUI repo).

Done :tiger2:

**Comment by fanquake** [comment]
Should we close this (for now)? This is blocked on cmake, as we aren't going to introduce a --with-qt6 option, or the other build system changes here. Looks like fb872edc1996527b6951899ff0127a6b815d0a41 could possibly be PR'd on it's own? Otherwise I'm not sure this is reviewable / that useful to keep open?

**Pull request closed without merging**